### PR TITLE
feat: switch assets based on address

### LINKED ISSUE #374: UX: Switch assets based on address/invoice input
When a Liquid address is entered/scanned, the frontend should switch the receiving symbol the L-BTC. Similar concept for Bitcoin addresses and lightning invoices/lnurls/sat domains

### PR DESCRIPTION
**Pull request by dni**
closes #374

switches onchain asset based on address paste

### CODE REVIEW

File: src/utils/compat.ts [hunk 1]
```diff
@@ -83,6 +86,33 @@ const decodeAddress = (asset: string, addr: string): DecodedAddress => {
...
+    });
+    liquidPrefixes.forEach((prefix: string) => {
+        if (address.startsWith(prefix)) {
+            asset = LBTC;
+        }
+    });
+    if (asset !== null) {
+        return asset;
+    }
+    // fallback to more costly decoding
```
Comment (michael1011) [inline src/utils/compat.ts:104]: Did you measure the performance of decoding the address?

File: src/utils/compat.ts [hunk 2]
```diff
@@ -26,7 +26,10 @@ import {
 import { Network as LiquidNetwork } from "liquidjs-lib/src/networks";
 
 import { network } from "../config";
-import { LBTC } from "../consts";
+import { BTC, LBTC } from "../consts";
+
+const btcPrefixes = ["bc", "tb", "bcrt"];
+const liquidPrefixes = ["lq", "ex", "tlq", "tex", "el", "ert"];
```
Comment (michael1011) [inline src/utils/compat.ts:32]: Don't hardcode those; if anything, get them from the configure network. Also those are only the b(l)ech32(m) ones?

File: src/utils/compat.ts [hunk 3]
```diff
@@ -83,6 +86,33 @@ const decodeAddress = (asset: string, addr: string): DecodedAddress => {
...
+    }
+    // fallback to more costly decoding
+    try {
+        decodeAddress(BTC, address);
+        return BTC;
+    } catch (e) {}
+    try {
+        decodeAddress(LBTC, address);
+        return LBTC;
+    } catch (e) {}
```
Comment (michael1011) [inline src/utils/compat.ts:112]: We need to decode the address anyways. Why don't we start decoding with the set network and then try all the others known to us?

- if address is valid for set network -> address valid
- address is valid for other network -> address valid & switch network
- address is not valid for any network -> address invalid

File: src/components/AddressInput.tsx [hunk 4]
```diff
@@ -1,32 +1,54 @@
...
         const inputValue = input.value.trim();
         const address = extractAddress(inputValue);
+        const invoice = extractInvoice(inputValue);
+
+        // auto switch direction based on invoice
+        if (isInvoice(invoice)) {
+            setAssetReceive(LN);
+            setAssetSend(asset());
+            setInvoice(invoice);
+            notify("success", t("switch_paste"));
```
Comment (michael1011) [inline src/components/AddressInput.tsx:37]: Not a fan of throwing a notification for every action you do

**Review by michael1011 — CHANGES_REQUESTED** [review]

File: src/components/AddressInput.tsx (hunk 4 above)
Comment (dni) [reply src/components/AddressInput.tsx:37]: i like it, and i talked with @kilrau about it.

File: src/utils/compat.ts (hunk 1 above)
Comment (dni) [reply src/utils/compat.ts:104]: i did not measure it, but noticed the browser hanging sometimes decoding the address

File: src/utils/compat.ts (hunk 3 above)
Comment (dni) [reply src/utils/compat.ts:112]: we  dont need to decode the address when pasting into the `InvoiceInput` and `setOnchainAddress` would decode it twice

**Comment by kilrau** [comment]
I think this branch needs a rebase, I have rootstock dangling in there 
image

File: src/i18n/i18n.ts [hunk 5]
```diff
@@ -707,6 +712,7 @@ const dict = {
         swap_in_progress: "此交换仍在进行中。",
         paste_invalid: "剪贴板包含无效字符或超出最大金额",
         email: "邮箱",
+        switch_paste: "根据粘贴的内容切换交换方向/资产",
```
Comment (michael1011) [inline src/i18n/i18n.ts:715]: @kilrau check the translations

**Review by michael1011 — APPROVED** [review]

File: src/i18n/i18n.ts (hunk 5 above)
Comment (kilrau) [reply src/i18n/i18n.ts:715]: LGTM

**Review by kilrau — CHANGES_REQUESTED** [review]
Notification of switch doesn't always show, e.g. here when switching back from lightning to liquid:
[Screencast from 03-20-2024 11:39:06 AM.webm](https://github.com/BoltzExchange/boltz-web-app/assets/19181985/5ffa23a6-cd60-41a5-bbc2-c2a8b274ebdc)

**Review by kilrau — APPROVED** [review]
LGTM

**Pull request merged**